src/libexpr/flake/flake.cc

```diff
@@ -229,8 +233,20 @@ static Flake getFlake(
 
     auto sInputs = state.symbols.create("inputs");
 
-    if (auto inputs = vInfo.attrs->get(sInputs))
+    if (auto inputs = vInfo.attrs->get(sInputs)) {
         flake.inputs = parseFlakeInputs(state, inputs->value, inputs->pos, flakeDir, lockRootPath);
+        if (originalRef.input.attrs.count("submodules") == 0) {
+            auto self = flake.inputs.find("self");
+            if (self != flake.inputs.end() && self->second.hasSubmodules) {
```

Comment:
This comparison is a little ad hoc. It should be described in more general terms. My understanding of the process is:
1. Nix makes a guess what the `self` input is
2. Nix checks that its guess matches the declared `self` input
3. Adjust

None of this is specific to git or submodules, so it should be done through virtual methods.

Other than that, I think lazy trees #6530 should make this specific parameter conceptually redundant, and enable it by default, as submodules should be fetched lazily.
However, that's not a reason not to have this mechanism, as we'll need it for at least a `git-crypt` parameter.

Reply:
A general mechanism (but also with `submodules` as the motivating case) is described and discussed in
 - #5312